[18.x]  WFLY-12144 Add support for configuring load-balancer with ranked routing

rhusar opened the pull request:
Jira
https://issues.jboss.org/browse/WFLY-12144

Proposal 
~~https://github.com/wildfly/wildfly-proposals/pull/150 merged now~~
https://github.com/wildfly/wildfly-proposals/pull/239 LB updates only

Requires
https://github.com/wildfly/wildfly/pull/12346 (subsystem version bump; included here)

preferably merged with
~~https://github.com/wildfly/wildfly/pull/11791 (server support for ranked) merged now~~

rhusar committed:
WFLY-12144 Add support for configuring load-balancer with ranked routing

rhusar committed:
WFLY-12173 Bump the Undertow subsystem model version and schema to version 10

pferraro commented on undertow/src/test/java/org/wildfly/extension/undertow/UndertowSubsystem11TestCase.java:
```diff
@@ -76,7 +76,7 @@ public void testRuntime() throws Exception {
     @Override
     protected AdditionalInitialization createAdditionalInitialization() {
-        return UndertowSubsystemTestCase.DEFAULT;
+        return UndertowSubsystem100TestCase.DEFAULT;
```
AbstractUndertowSubsystemTestCase.DEFAULT

pferraro commented on undertow/src/test/java/org/wildfly/extension/undertow/UndertowSubsystem12TestCase.java:
```diff
@@ -65,14 +65,14 @@ public void testSubsystem() throws Exception {
     public void testRuntime() throws Exception {
         setProperty();
-        KernelServicesBuilder builder = createKernelServicesBuilder(UndertowSubsystemTestCase.RUNTIME)
+        KernelServicesBuilder builder = createKernelServicesBuilder(UndertowSubsystem100TestCase.RUNTIME)
```
AbstractUndertowSubsystemTestCase.RUNTIME

pferraro commented on undertow/src/test/java/org/wildfly/extension/undertow/UndertowSubsystem12TestCase.java:
```diff
@@ -65,14 +65,14 @@ public void testSubsystem() throws Exception {
     @Override
     protected AdditionalInitialization createAdditionalInitialization() {
-        return UndertowSubsystemTestCase.DEFAULT;
+        return UndertowSubsystem100TestCase.DEFAULT;
```
AbstractUndertowSubsystemTestCase.DEFAULT

pferraro commented on undertow/src/test/java/org/wildfly/extension/undertow/UndertowSubsystem20TestCase.java:
```diff
@@ -67,7 +67,7 @@ protected KernelServices standardSubsystemTest(String configId, boolean compareX
     public void testRuntime() throws Exception {
         setProperty();
-        KernelServicesBuilder builder = createKernelServicesBuilder(UndertowSubsystemTestCase.RUNTIME)
+        KernelServicesBuilder builder = createKernelServicesBuilder(UndertowSubsystem100TestCase.RUNTIME)
```
AbstractUndertowSubsystemTestCase.RUNTIME

pferraro commented on undertow/src/test/java/org/wildfly/extension/undertow/UndertowSubsystem20TestCase.java:
```diff
@@ -77,7 +77,7 @@ public void testRuntime() throws Exception {
     @Override
     protected AdditionalInitialization createAdditionalInitialization() {
-        return UndertowSubsystemTestCase.DEFAULT;
+        return UndertowSubsystem100TestCase.DEFAULT;
```
AbstractUndertowSubsystemTestCase.DEFAULT

pferraro commented on undertow/src/main/java/org/wildfly/extension/undertow/filters/ModClusterDefinition.java:
```diff
@@ -325,26 +335,72 @@ public HttpHandler createHttpHandler(Predicate predicate, ModelNode model, HttpH
+                .addRequiredSingletonChildren(SingleAffinityResourceDefinition.PATH)
+                ;
+
+        for (AttributeDefinition attribute : ATTRIBUTES) {
```
Create an enumeration for the attributes, and eliminate the ATTRIBUTES variable.

pferraro commented on undertow/src/main/java/org/wildfly/extension/undertow/filters/ModClusterDefinition.java:
```diff
@@ -325,26 +335,72 @@ public HttpHandler createHttpHandler(Predicate predicate, ModelNode model, HttpH
+        }
+    }
+
+    static class ModClusterAddStepHandler extends AddStepHandler {
```
There are better ways to provide a custom Resource implementation without needing to extend AddStepHandler, but they will require refactoring ModClusterResource itself.  But we can address this separately.

pferraro commented on undertow/src/test/java/org/wildfly/extension/undertow/UndertowSubsystem10TestCase.java:
```diff
@@ -68,15 +68,15 @@ protected KernelServices standardSubsystemTest(String configId, boolean compareX
 
         setProperty();
-        KernelServicesBuilder builder = createKernelServicesBuilder(UndertowSubsystemTestCase.RUNTIME)
+        KernelServicesBuilder builder = createKernelServicesBuilder(UndertowSubsystem100TestCase.RUNTIME)
```
Access static fields directly, e.g. AbstractUndertowSubsystemTestCase.RUNTIME

pferraro commented on undertow/src/test/java/org/wildfly/extension/undertow/UndertowSubsystem10TestCase.java:
```diff
@@ -68,15 +68,15 @@ protected KernelServices standardSubsystemTest(String configId, boolean compareX
     @Override
     protected AdditionalInitialization createAdditionalInitialization() {
-        return UndertowSubsystemTestCase.DEFAULT;
+        return UndertowSubsystem100TestCase.DEFAULT;
```
Access static fields directly, e.g. AbstractUndertowSubsystemTestCase.DEFAULT

pferraro commented on undertow/src/main/java/org/wildfly/extension/undertow/UndertowSubsystemParser_10_0.java:
```diff
@@ -0,0 +1,441 @@
+                                        .addAttributes(ErrorPageDefinition.CODE, ErrorPageDefinition.PATH)
+                        ).addChild(
+                                builder(ModClusterDefinition.INSTANCE.getPathElement())
+                                .addAttributes(
```
If you convert these attributes to an enumeration, you can replace all of this with:
new AttributeXMLBuilderOperator(ModClusterDefinition.Attribute.class).apply(builder)

pferraro commented on undertow/src/main/java/org/wildfly/extension/undertow/UndertowTransformers.java:
```diff
@@ -112,6 +115,13 @@ private static void registerTransformers_EAP_7_2_0(ResourceTransformationDescrip
+        ResourceTransformationDescriptionBuilder builder = subsystemBuilder.addChildResource(UndertowExtension.PATH_FILTERS)
+                .addChildResource(PathElement.pathElement(Constants.MOD_CLUSTER));
+
+        builder.discardChildResource(NoAffinityResourceDefinition.PATH);
```
The legacy behavior corresponds to SingleAffinityResourceDefinition.PATH, thus I think we want to use:
builder.rejectChildResource(NoAffinityResourceDefinition.PATH)

rhusar replied:
Oops, fixed.

pferraro commented on undertow/src/main/java/org/wildfly/extension/undertow/UndertowTransformers.java:
```diff
@@ -112,6 +115,13 @@ private static void registerTransformers_EAP_7_2_0(ResourceTransformationDescrip
+
+        builder.discardChildResource(NoAffinityResourceDefinition.PATH);
+        builder.discardChildResource(SingleAffinityResourceDefinition.PATH);
+        builder.discardChildResource(RankedAffinityResourceDefinition.PATH);
```
builder.rejectChildResource(RankedAffinityResourceDefinition.PATH)

rhusar replied:
Oops, fixed.

pferraro commented on undertow/src/main/java/org/wildfly/extension/undertow/filters/AffinityResourceDefinition.java:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    AffinityResourceDefinition(PathElement path) {
+        super(path, UndertowExtension.getResolver(Constants.HANDLER, Constants.AFFINITY, path.getValue()), new ReloadRequiredAddStepHandler(), new ReloadRequiredRemoveStepHandler());
```
Unless I'm mistaken, you don't need to reload the whole server, just restart the services of the parent resource.
You can leverage RestartParentResourceRegistration for this.

pferraro replied:
Nevermind - while simply restarting the parent service is preferable to reloading the entire server, this isn't easily done without significant refactoring, since ModClusterService doesn't properly encapsulate service installation.

pferraro commented on undertow/src/main/java/org/wildfly/extension/undertow/filters/RankedAffinityResourceDefinition.java:
```diff
@@ -0,0 +1,76 @@
+    public static final PathElement PATH = pathElement(Constants.RANKED);
+
+    public enum Attribute implements org.jboss.as.clustering.controller.Attribute {
+        DELIMITER(Constants.DELIMITER, ModelType.STRING, null),
```
Why not have a default value?

rhusar replied:
We can... defaulting to "." now.

pferraro commented on undertow/src/main/java/org/wildfly/extension/undertow/filters/ModClusterDefinition.java:
```diff
@@ -325,26 +335,72 @@ public HttpHandler createHttpHandler(Predicate predicate, ModelNode model, HttpH
+            });
+        }
+
+        ModClusterResourceServiceHandler handler = new ModClusterResourceServiceHandler();
```
I don't think you need a custom ResourceServiceHandler implementation.  Just use ParentResourceServiceHandler.
This will allow you to leverage the RestartParentResource*Handlers.

pferraro replied:
Nevermind - this isn't easily done without significant refactoring, since ModClusterService doesn't properly encapsulate service installation.

rhusar commented:
Addressed the issues for now, the rest requires substantial refactor (custom resources, custom attribute definitions OptionAttributeDefinition, etc) which I can try to get in, but I don't see as holding this PR.

rhusar commented:
Prested by QE's @tommaso-borgato

pferraro commented on undertow/src/main/java/org/wildfly/extension/undertow/filters/AffinityResourceDefinition.java:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    AffinityResourceDefinition(PathElement path) {
+        super(path, UndertowExtension.getResolver(Constants.HANDLER, Constants.AFFINITY, path.getValue()), new ReloadRequiredAddStepHandler(), new ReloadRequiredRemoveStepHandler());
```
Don't use this super-constructor. Your subclasses already register different add/remove step handlers.

rhusar replied:
Done.

pferraro commented on undertow/src/main/java/org/wildfly/extension/undertow/filters/AffinityResourceDefinition.java:
```diff
@@ -0,0 +1,47 @@
+ *
+ * @author Radoslav Husar
+ */
+public abstract class AffinityResourceDefinition extends SimpleResourceDefinition {
```
I suggest extending ChildResourceDefinition instead, since you are registering everything in a single method.

rhusar replied:
DOne.

pferraro commented on undertow/src/main/java/org/wildfly/extension/undertow/filters/ModClusterDefinition.java:
```diff
@@ -325,26 +335,72 @@ public HttpHandler createHttpHandler(Predicate predicate, ModelNode model, HttpH
     }
 
+    @Override
+    public void registerOperations(ManagementResourceRegistration parent) {
```
Since your registering more than the operations here, I would suggest using a single registration method.

rhusar replied:
OK

pferraro commented on undertow/src/main/java/org/wildfly/extension/undertow/filters/ModClusterDefinition.java:
```diff
@@ -322,29 +333,81 @@ private ModClusterDefinition() {
+        new SingleAffinityResourceDefinition().register(parent);
+        new RankedAffinityResourceDefinition().register(parent);
+
+        parent.registerSubModel(ModClusterBalancerDefinition.INSTANCE);
```
These are child resources, no?  Should you register them using the MRR of the ModClusterResourceDefinition?

rhusar replied:
No.

But as discussed on the call I have reworked this.

bstansberry commented:
@fl4via please have a look to be sure the added management API (i.e. new resources, attributes) is ok with you.

bstansberry commented:
I put the missing-reqs label on this while the https://github.com/wildfly/wildfly-proposals/pull/150 document gets updated to cover the requirements for this bit.

rhusar committed:
WFLY-12144 Update documentation

rhusar commented:
Given #11791 has been merged now, rebased this branch with new integration tests, documentation and updated proposal.

rhusar commented on clustering/common/src/main/java/org/jboss/as/clustering/controller/ResourceDescriptor.java:
```diff
@@ -129,6 +129,18 @@ public ResourceDescriptor addAttributes(Iterable<? extends Attribute> attributes
 
+    public ResourceDescriptor addAttributes(Collection<AttributeDefinition> attributes) {
+        for (AttributeDefinition attribute : attributes) {
+            this.addAttributes(new Attribute() {
```
Oops. Fixed.

rhusar committed:
WFLY-12144 Add integration test for ranked affinity/routing

rhusar commented:
retest this please

rhusar commented:
@jamezp Please let me know if anything is holding this up I can do.
